File: IRC_client.java

import java.util.*;
import java.io.InputStream;
import java.util.Scanner;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class IRC_client {
    static ArrayList history_array = new ArrayList();
    static BufferedReader br=null;
    static String nick = "";
    public static void deamon(BufferedReader br, String nick){
        while(true){
            String nickname = "";
            String message = "";
            String resp="";
            try{
                  resp = br.readLine();
            }catch(Exception e){
                 System.out.println("Error, broken pipe or something...");
                 System.exit(-1);
            }
            if (resp!=null){
                if (resp!=""){
                    if (resp.contains("PRIVMSG")){
                           nickname = resp.replace("!", ":").split(":")[1];
                           message = resp.split(":")[2];
                           resp = nickname+":"+message;          
                     }
                     history_array.add(resp);
                     for (int i=0; i<history_array.size(); i++){
                         System.out.println(history_array.get(i));
                     }
                     System.out.print(nick+">> ");
                    
                }
            
            }
        }
    }
    public static String raw_input(String text){
        Scanner scanner = new Scanner( System.in );
        System.out.print(text);
        String irc = scanner.nextLine();
        return irc;

    }
    public static void main(String[] args) {
        // Prints "Hello, World" to the terminal window.
        System.out.println("Welcome to Greg's IRC client.");
        String irc = raw_input("Enter IRC server ip and port:");
        System.out.print("Connecting to "+irc+"... ");
        Socket sock2 = null;

        PrintWriter out=null;
        try{
        
             String host = irc.split(":")[0];
             int port = Integer.parseInt(irc.split(":")[1]);
             sock2 = new Socket(host, port);
             br = new BufferedReader(new InputStreamReader(sock2.getInputStream()));
             out = new PrintWriter(sock2.getOutputStream(), true);
        }catch(Exception e){
            System.out.println("FAILED");
            System.out.println("Cause:");
            System.out.println(e);
            System.exit(-1);
        }
        System.out.println("OK");
        nick = raw_input("Enter your nick:");
        String full_name = raw_input("Enter your full name:");
        out.println("NICK "+nick+"\r\n");
        out.println("USER "+nick+" * * :"+full_name+"\r\n");
        String resp="";
        try{
            
            while(true){
                String line = br.readLine();
                if (line!=null){
                    resp = resp+"\n"+line;
                
                    System.out.println(line);
                
                    if(resp.toString().contains("PING :")){
                       out.println("PONG :"+resp.split("PING :")[1]+"\r\n");
                       break;
                    }
                }
            }
        }catch(Exception e){
            System.err.println(e);
            System.exit(-1);
        }
         Thread t1 = new Thread(new Runnable() {
         public void run()
         {
             deamon(br, nick);
         }});
         t1.start();
         String chatroom = "";
         while(true){
               for (int i=0; i<history_array.size(); i++){
                    System.out.println(history_array.get(i));
               }
               String command = raw_input(nick+">> ");
               command = command.replace("\n", "");
               command = command.replace("/join", "/JOIN");
               if (command.contains("/JOIN")){
                   System.out.println("Joining "+command.split("JOIN")[1].replace(" ","")+"...");
                   chatroom=command.split("JOIN")[1].replace(" ","");
                   command=command.replace("/","");
               }else{
                   if (!(command.contains("/"))){
                       command = "PRIVMSG "+chatroom+" :"+command;
                   }else{
                       command = command.replace("/", "");
                   }
               }
               if (command != "PRIVMSG "){
                     out.println(command +"\r\n");
                     if (command.contains("PRIVMSG")){
                         command = nick+":"+command.split(":")[1];
                     }
                     history_array.add(command);
               
               
               
               }
         
         
         
         
         }
        
        

    }

}

File: IRC-client.py
import socket
import sys
import threading
history_array = []
for i in range(0, 42):
    history_array.append("")
def deamon(sock2, nick):
    global history_array
    while True:
        resp = sock2.recv(2048)
        if not resp=="":
             if "PRIVMSG" in resp:
                 nickname = resp.replace("!", ":").split(":")[1]
                 message = resp.split(":")[2]
                 resp = nickname+":"+message
             history_array.append(resp)
             for line in history_array:
                print line
             sys.stdout.write(nick+">> ")
             sys.stdout.flush()
print "Welcome to Greg's IRC client."
irc = str(raw_input( "Enter IRC server ip and port:"))
sys.stdout.write("Connecting to "+irc+"... ")
sys.stdout.flush()


try:
    host = irc.split(":")[0]
    port = irc.split(":")[1]
    sock2 = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                
# Connect the socket to the port where the server is listening
    server_address2 = (host.replace("\n","").replace("\n", ""), int(port))
    sock2.connect(server_address2)

except Exception as e:
    print "FAILED"
    print str(e)
    quit()
print "OK"
#print 'connecting to %s port %s' % server_address2
#print str(server_address2)

nick = str(raw_input("Enter your nick:"))
full_name = str(raw_input("Enter your full name:"))
sock2.sendall("NICK "+nick+"\r\n")
sock2.sendall("USER "+nick+" * * :"+full_name+"\r\n")
resp = sock2.recv(2048)
print resp
chatroom = ""
pingPong = sock2.recv(512)
print "recv 2", pingPong
pingPong = pingPong[5:]
sock2.sendall('PONG %s\r\n' % pingPong)

t1 = threading.Thread(target=deamon, args=[sock2, nick])
t1.start()

while True:
    for line in history_array:
                print line
    command = str(raw_input(nick+">> "))
    command = command.replace("\n", "")
    command = command.replace("/join", "/JOIN")
    if "/JOIN" in command:
        print "Joining "+command.split("JOIN")[1].replace(" ","")+"..."
        chatroom=command.split("JOIN")[1].replace(" ","")
        command=command.replace("/","")
    else:
        print "chatroom:"+chatroom
        if not "/" in command:
            command = "PRIVMSG "+chatroom+" :"+command
            print "Command:"+command
            #history_array.append(command)
        else:
            command = command.replace("/", "")
    if not "PRIVMSG " == command:
         
        sock2.sendall(command +"\r\n")
        if "PRIVMSG" in command:
            command = nick+":"+command.split(":")[1]
        history_array.append(command)
    #resp = sock2.recv(2048)
    #print resp

